### PR TITLE
Initial storage implementation for native blockifier

### PR DESCRIPTION
**Pull request by giladchase**
- Wrap papyrus_storage with a dedicated class to hide the implementation.
- Implement initialization API.
- Add error handling.

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 6 unresolved discussions (waiting on @alonh5 and @giladchase)

___
*[`crates/native_blockifier/Cargo.toml` line 17 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHWok8A7PlujFB_ax1:-NNHWok8A7PlujFB_ax2:banifof) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/Cargo.toml#L17)):*
> ```TOML
> [dependencies]
> pyo3 = { version = "0.17.3", features = ["extension-module"] }
> papyrus_storage = { git = "https://github.com/starkware-libs/papyrus", features = [
> ```

I keep forgetting; what does `features = [ "testing", ] }` mean? Does it bring the whole crate + all that's under `[#test]` in it?
___
*[`crates/native_blockifier/src/lib.rs` line 1 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXSXr0k9gi-2CC-my:-NNHXSXr0k9gi-2CC-mz:b-i5zta7) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L1)):*
> ```XML
> use papyrus_storage::db::DbConfig;
> ```

Can we use the full module name, so it's clear it's the Papyrus storage? (as you did in `PapyrusStateReader`).



_Code quote:_
```XML
use papyrus_storage::db::DbConfig;
use papyrus_storage::{open_storage, StorageReader, StorageWriter};
```
___
*[`crates/native_blockifier/src/lib.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXKVH92YjKdY4T2sL:-NNHXKVH92YjKdY4T2sM:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L10)):*
> ```XML
> 
> #[pymodule]
> fn native_blockifier(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
> ```





_Suggestion:_
```XML
py_module
```
___
*[`crates/native_blockifier/src/lib.rs` line 13 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXHy31d2kVVBHUXRY:-NNHXHy31d2kVVBHUXRZ:byfk56p) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L13)):*
> ```XML
>     m.add_class::<Storage>()?;
> 
>     m.add_function(wrap_pyfunction!(hello_world, m)?)?;
> ```

Do we want to remove these?



_Code quote:_
```XML
    m.add_function(wrap_pyfunction!(hello_world, m)?)?;
    m.add_function(wrap_pyfunction!(test_ret_value, m)?)?;
```
___
*[`crates/native_blockifier/src/lib.rs` line 48 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXud-AroGEOicYPjP:-NNHXud-AroGEOicYPjQ:b-sqhg0h) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L48)):*
> ```XML
> }
> 
> #[pyfunction]
> ```

Perhaps `new`, or `From Path`?
___
*[`crates/native_blockifier/src/lib.rs` line 70 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXC1R80VaIVTL6B16:-NNHXC1R80VaIVTL6B17:b-u9ygqo) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L70)):*
> ```XML
> 
> impl From<NativeBlockifierError> for PyErr {
>     fn from(err: NativeBlockifierError) -> PyErr {
> ```

Unless it's a Rust convention.



_Suggestion:_
```XML
error
```

**Review by elintul — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 7 unresolved discussions (waiting on @alonh5 and @giladchase)

*[a discussion](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHjWLX3rHLL0spubu5:-NNHjWLY0BnJKVfHJgTU:b-ex8xg) (no related file):*
Alon is also adding code here.
Let's move all `StateDiff`-related code to `py_state_diff.rs`.
Another option is implementing all Python-related converters to `py_objects.rs`; depends on the amount of code.
We can start with it and split afterwards. @alonh5 

___

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 9 unresolved discussions (waiting on @giladchase)

___
*[`crates/native_blockifier/src/lib.rs` line 66 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf61VO9fRdzTMomF_d:-NNf61VO9fRdzTMomF_e:b-ttztnx) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L66)):*
> ```XML
>     StarknetApiError(#[from] StarknetApiError),
>     #[error(transparent)]
>     Pyo3Error(#[from] PyErr),
> ```

Alphabetize :)
___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNf3QB09NcBZqBQQ9K4:b-8og5um) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
> ```XML
> impl From<NativeBlockifierError> for PyErr {
>     fn from(err: NativeBlockifierError) -> PyErr {
>         PyOSError::new_err(err.to_string())
> ```

If the error is `NativeBlockifierError::Pyo3Error`, you can return the internal error rather than wrapping it twice.

**Review by giladchase — COMMENTED** [review]
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 6 unresolved discussions (waiting on @alonh5 and @elintul)

*[a discussion](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHjWLX3rHLL0spubu5:-NNjaDsK4G_dWdKKsOqr:bcfqzv6) (no related file):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Alon is also adding code here.
Let's move all `StateDiff`-related code to `py_state_diff.rs`.
Another option is implementing all Python-related converters to `py_objects.rs`; depends on the amount of code.
We can start with it and split afterwards. @alonh5 
</blockquote></details>

OK, will do this after addressing your comments below to not turn them stale.

___
___
*[`crates/native_blockifier/Cargo.toml` line 17 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHWok8A7PlujFB_ax1:-NNVABn2BjdII7VX58qm:b-g2ra8j) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/Cargo.toml#L17)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I keep forgetting; what does `features = [ "testing", ] }` mean? Does it bring the whole crate + all that's under `[#test]` in it?
</blockquote></details>

Ya it's a confusing name... 
It's not a builtin, just a name papyrus has given to some things they want to conditionally compiled (test related stuff).

No relation to `#[test]`.
___
*[`crates/native_blockifier/src/lib.rs` line 13 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXHy31d2kVVBHUXRY:-NNjeWoSB6AKzlxA23Ov:ba6451j) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L13)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Do we want to remove these?
</blockquote></details>

Sure.
Also fixed the build-wheels script, since it used to check validity with the hello-world func.
___
*[`crates/native_blockifier/src/lib.rs` line 48 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNHXud-AroGEOicYPjP:-NNjfIb1AUfOCjPL06gs:b-jzvyif) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L48)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Perhaps `new`, or `From Path`?
</blockquote></details>

`new` is good.
I prefer it over `from` since this should be the only valid way of initializing storage from Python.

This has the added benefit of hiding the default `Storage` constructor from `Python` (Python has no business constructing `Storage` with a reader/writer).

This happens because the `#[new]` attr has special meaning in Pyo3: this is now the only way to initialize the `Storage` struct in Python; and _replaces_ the default constructor from python (so in Python, we initialize with `Storage(path)`, rather than `Storage.new(path)`).
___
*[`crates/native_blockifier/src/lib.rs` line 66 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf61VO9fRdzTMomF_d:-NNjhz-Z5MpMHOgqzUgb:b-xfk72a) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L66)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Alphabetize :)
</blockquote></details>

:laughing: 
Done.
___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNjiQmMEjUmIxG7Zkmp:bfgzfj6) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

If the error is `NativeBlockifierError::Pyo3Error`, you can return the internal error rather than wrapping it twice.
</blockquote></details>

Ya, i think it was leftover from a refactor.
Removed.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 3 of 5 files reviewed, 6 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNjs4Ws34iN2-XHjEec:b-tdtr2f) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Ya, i think it was leftover from a refactor.
Removed.
</blockquote></details>

OK i remember now why i put it there: in the other PR I have a function that may raise both a Pyo3 error and a `papyrus_storage` error.
It's subtle...
Added it to the second PR though, since it makes more sense to put it there.



_Code quote:_
```XML
    #[error(transparent)]
    Pyo3Error(#[from] PyErr),
```

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNk0ADE-hWQUROHlyzx:b-d8ocz9) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

OK i remember now why i put it there: in the other PR I have a function that may raise both a Pyo3 error and a `papyrus_storage` error.
It's subtle...
Added it to the second PR though, since it makes more sense to put it there.

</blockquote></details>

It's fine and probably necessary to have the `Pyo3Error(#[from] PyErr)` in `NativeBlockifierError`. I meant the `From<NativeBlockifierError> for PyErr` can return the inner `PyErr` in that case, and not wrap it again so yo don't get `PyErr(NativeBlockifierError(PyErr(...)))`.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNlOyI1CJ-uGHN3U2Gp:b-czwzc8) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

It's fine and probably necessary to have the `Pyo3Error(#[from] PyErr)` in `NativeBlockifierError`. I meant the `From<NativeBlockifierError> for PyErr` can return the inner `PyErr` in that case, and not wrap it again so yo don't get `PyErr(NativeBlockifierError(PyErr(...)))`.
</blockquote></details>

IUUC `to_string()` doesn't take the inner `NativeBlockifierError(PyErr(` into account, only the actual error message.
This is because `transparent` forwards the `display` method from the original error "as is" (which is called when calling `to_string()` _explicitly_)

That is,  `PyOSError::new_err(err.to_string())` is identical to the "original" `PyErr`.

Made an example just to convince myself of this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=50f2a857696646a74d3c54d987a9a4f0

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 5 unresolved discussions (waiting on @alonh5, @elintul, and @giladchase)

___
*[`crates/native_blockifier/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNluyuR49hMLuanBkt6:-NNluyuR49hMLuanBkt7:b-cm9ff2) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/Cargo.toml#L20)):*
> ```TOML
>     "testing",
> ] }
> starknet_api = { git = "https://github.com/starkware-libs/starknet-api", rev = "0a11ad3" }
> ```

Can we delete this and take latest?



_Code quote:_
```TOML
rev = "0a11ad3"
```
___
*[`crates/native_blockifier/src/lib.rs` line 18 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNlvBy-6Wgs27e1vHQg:-NNlvBy-6Wgs27e1vHQh:b-5qg64i) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/src/lib.rs#L18)):*
> ```XML
> #[pyclass]
> pub struct Storage {
>     pub reader: papyrus_storage::StorageReader,
> ```

Where is it imported?



_Code quote:_
```XML
papyrus_storage
```
___
*[`crates/native_blockifier/src/lib.rs` line 41 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNlvp5n81KrK7SiGYH1:-NNlvp5n81KrK7SiGYH2:b-x70ep2) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/src/lib.rs#L41)):*
> ```XML
>     let (reader, writer) = papyrus_storage::test_utils::get_test_storage();
> 
>     Storage { reader, writer }
> ```

Remove blank line above.



_Code quote:_
```XML
Storage { reader, writer }
```

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 4 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/src/lib.rs` line 18 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNlvBy-6Wgs27e1vHQg:-NNoZw2ICBafmYhOJn-1:b-vh4h6x) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/src/lib.rs#L18)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Where is it imported?
</blockquote></details>

[No need to import size-1 path](https://rust-lang.github.io/rust-clippy/master/#single_component_path_imports).

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 2 of 5 files reviewed, 4 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNluyuR49hMLuanBkt6:-NNofAS5EOU1jY6jzLr4:bjxh6lu) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/Cargo.toml#L20)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Can we delete this and take latest?
</blockquote></details>

Oooh good point :nerd_face: 
Done.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 2 of 5 files reviewed, 4 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNluyuR49hMLuanBkt6:-NNpgF_x61YZ2jFJCAWF:bm28nup) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/Cargo.toml#L20)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Oooh good point :nerd_face: 
Done.
</blockquote></details>

OK i think i need to put it back :(
It has to match the `rev` in `blockifier` and `papyrus_storage`, since we are both using the same objects from starknet_api (and cargo won't have `StarkFelt`s originating from two different versions of the same crate at once).

**Review by elintul — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262#-:-NNpjxR510xG3ebMh951:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @alonh5 and @giladchase)

___
*[`crates/native_blockifier/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNluyuR49hMLuanBkt6:-NNpjibL0GQxY1p3Bs84:b-5xi4s7) ([raw file](https://github.com/starkware-libs/blockifier/blob/e08d0474634e1ad4cb1c9d88850677b5b3e6236d/crates/native_blockifier/Cargo.toml#L20)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

OK i think i need to put it back :(
It has to match the `rev` in `blockifier` and `papyrus_storage`, since we are both using the same objects from starknet_api (and cargo won't have `StarkFelt`s originating from two different versions of the same crate at once).
</blockquote></details>

Then can you add a short comment on that?
___
*[`crates/native_blockifier/src/lib.rs` line 14 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNpjomS6EiWYXy8AEMD:-NNpjomS6EiWYXy8AEME:b3rltk6) ([raw file](https://github.com/starkware-libs/blockifier/blob/acded52e1bcac2b804903350ee036c22589e634d/crates/native_blockifier/src/lib.rs#L14)):*
> ```XML
>     py_module.add_function(wrap_pyfunction!(execute_tx, py_module)?)?;
> 
>     py_module.add_class::<Storage>()?;
> ```

Move to be first: classes, then methods; both alphabetized.



_Code quote:_
```XML
py_module.add_class::<Storage>()?;
```

**Review by elintul — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 2 unresolved discussions (waiting on @alonh5 and @giladchase)

___
*[`crates/native_blockifier/Cargo.toml` line 22 at r5](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNpoWb6ASzh-U6iJRPs:-NNpoWb6ASzh-U6iJRPt:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/db9bccc4b80495db8fb6d26502b24e222e445c07/crates/native_blockifier/Cargo.toml#L22)):*
> ```TOML
> ] }
> pyo3 = { version = "0.17.3", features = ["extension-module", "num-bigint"] }
> # We need this rev to match the one `blockifier` and `papyrus_storage` use.
> ```





_Suggestion:_
```TOML
to be the same one both
```
___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNpo_zz26qMJoE9xY3o:b-sfa4cr) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

IUUC `to_string()` doesn't take the inner `NativeBlockifierError(PyErr(` into account, only the actual error message.
This is because `transparent` forwards the `display` method from the original error "as is" (which is called when calling `to_string()` _explicitly_)

That is,  `PyOSError::new_err(err.to_string())` is identical to the "original" `PyErr`.

Made an example just to convince myself of this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=50f2a857696646a74d3c54d987a9a4f0
</blockquote></details>

TL;DR, please?

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @alonh5 and @elintul)

___
*[`crates/native_blockifier/Cargo.toml` line 22 at r5](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNpoWb6ASzh-U6iJRPs:-NNqI2fr6tHzJltO8NAU:bmn4et4) ([raw file](https://github.com/starkware-libs/blockifier/blob/db9bccc4b80495db8fb6d26502b24e222e445c07/crates/native_blockifier/Cargo.toml#L22)):*
> ```TOML
> ] }
> pyo3 = { version = "0.17.3", features = ["extension-module", "num-bigint"] }
> # We need this rev to match the one `blockifier` and `papyrus_storage` use.
> ```

changed to `both`->`as in both` (recalling tat `native_blockifier` and `blockifier` have separate `cargo.toml`s)
___
*[`crates/native_blockifier/src/lib.rs` line 71 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNf3QB09NcBZqBQQ9K3:-NNqGAet2rLIzA1gRaeb:bjbk313) ([raw file](https://github.com/starkware-libs/blockifier/blob/7b2ceff25d902b41360c207c8a0fad428ba13563/crates/native_blockifier/src/lib.rs#L71)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

TL;DR, please?
</blockquote></details>

For technical reasons, we need `NativeBlockifierError` to wrap `PyErr` (via "transparent"), but must also be able to _convert_ `NativeBlockifierError` into `PyErr`.

There are two ways of doing this; I reverted mine since Alon's way is (IMO) clearer.

**Review by elintul — APPROVED** [review]
> Reviewed 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 2 unresolved discussions (waiting on @alonh5 and @giladchase)

___
*[`crates/native_blockifier/Cargo.toml` line 22 at r5](https://reviewable.io/reviews/starkware-libs/blockifier/262#-NNpoWb6ASzh-U6iJRPs:-NNqYd_R3ZPini8XNXo4:b-6yv6vt) ([raw file](https://github.com/starkware-libs/blockifier/blob/db9bccc4b80495db8fb6d26502b24e222e445c07/crates/native_blockifier/Cargo.toml#L22)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

changed to `both`->`as in both` (recalling tat `native_blockifier` and `blockifier` have separate `cargo.toml`s)
</blockquote></details>

Remove `on`?

**Review by elintul — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/262)* status: all files reviewed, 1 unresolved discussion (waiting on @alonh5)
